pkg/operator/controllers/pullsecret/pullsecret_controller.go

```diff
@@ -162,3 +120,140 @@ func (r *PullSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Named(controllers.PullSecretControllerName).
 		Complete(r)
 }
+
+// fixAndUpdateGlobalPullSecret checks the state of the pull secrets, in case of missing or broken ARO pull secret
+// it replaces it with working one from controller Secret
+func (r *PullSecretReconciler) fixAndUpdateGlobalPullSecret(ctx context.Context, operatorSecret, userSecret *corev1.Secret) (err error) {
+	if operatorSecret == nil {
+		return errors.New("nil operator secret, cannot verify userData integrity")
+	}
+
+	var secret *corev1.Secret
+	create := userSecret == nil
+	remove := false
+
+	// userSecret can happen to have broken type, which means it have to be recreated
+	// with proper type
+	if userSecret != nil && (userSecret.Type != corev1.SecretTypeDockerConfigJson || userSecret.Data == nil) {
+		create = true
+		remove = true
+	}
+
+	if create {
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      pullSecretName.Name,
+				Namespace: pullSecretName.Namespace,
+			},
+			Type: corev1.SecretTypeDockerConfigJson,
+		}
+		secret.Data = make(map[string][]byte)
+	} else {
+		secret = userSecret.DeepCopy()
+		if !json.Valid(secret.Data[corev1.DockerConfigJsonKey]) {
+			delete(secret.Data, corev1.DockerConfigJsonKey)
+		}
+	}
+
+	fixedData, update, err := pullsecret.Merge(string(secret.Data[corev1.DockerConfigJsonKey]), string(operatorSecret.Data[corev1.DockerConfigJsonKey]))
+	if err != nil {
+		return err
+	}
+
+	if !create && !update {
+		return nil
+	}
+
+	if remove {
+		// unfortunately the type field is immutable.
+		err := r.kubernetescli.CoreV1().Secrets(secret.Namespace).Delete(ctx, secret.Name, metav1.DeleteOptions{})
+		if err != nil && !kerrors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	secret.Data[corev1.DockerConfigJsonKey] = []byte(fixedData)
+
+	if create {
+		_, err := r.kubernetescli.CoreV1().Secrets(secret.Namespace).Create(ctx, secret, metav1.CreateOptions{})
+
+		return err
+	}
+
+	_, err = r.kubernetescli.CoreV1().Secrets(secret.Namespace).Update(ctx, secret, metav1.UpdateOptions{})
+
+	return err
+}
+
+func (r *PullSecretReconciler) buildRedHatKeyCondition(secret *corev1.Secret) *status.Condition {
+	// parse keys and validate JSON
+	parsedKeys, err := pullsecret.UnmarshalSecretData(secret)
+	foundKeys := []string{}
+	failed := false
+	if err != nil {
+		r.log.Info("pull secret is not valid json - recreating")
+		failed = true
+	} else {
+		foundKeys = r.checkRHRegistryKey(parsedKeys)
+	}
+
+	keyCondition := r.keyCondition(failed, foundKeys)
+	return keyCondition
+}
+
+// checkRHRegistryKey checks whether the rhRegistry key:
+//   - redhat.registry.io
+//   - cloud.redhat.com
+// is present in the pullSecret
+func (r *PullSecretReconciler) checkRHRegistryKey(psData map[string]string) (foundKeys []string) {
+	if psData == nil {
+		return foundKeys
+	}
+
+	for _, rhKey := range rhKeys {
+		for k, v := range psData {
+			if k == rhKey && len(v) > 0 {
```

Comment:
O(N^2) although admittedly N is always small.
Should be something like `if _, found := psData[rhKey]; found {`